File: EPI/9.10/typescript/9.10.ts
export default class CircularQueue<T> {
    private _size = 0;
    private d = 0;
    private e = 0;

    private array: (T | undefined)[];

    constructor(capacity: number) {
        this.array = new Array(capacity);
    }

    get size(): number {
        return this._size;
    }

    enqueue(element: T): void {
        if (this._size === this.array.length) {
            this.grow();
        }

        this.array[this.e] = element;
        this.e = (this.e + 1) % this.array.length;
        this._size += 1;
    }

    dequeue(): T | undefined {
        if (this._size === 0) {
            return undefined;
        }

        const element = this.array[this.d];
        this.array[this.d] = undefined; // Let the element be GC'ed.
        this.d = (this.d + 1) % this.array.length;
        this._size -= 1;
        return element;
    }

    private grow(): void {
        const newArray = new Array(this.array.length * 2);

        for (let i = 0; i < this.array.length; i += 1) {
            newArray[i] = this.array[(this.d + i) % this.array.length];
        }

        this.d = 0;
        this.e = this.array.length;
        this.array = newArray;
    }
}

File: EPI/Stacks and Queues/Implement a circular queue/swift/9.10.swift
public struct Queue<T> {
    var array = [T]()
    var firstIndex = 0

    public var count = 0
    public var capacity: Int

    public init(capacity: Int) {
        self.array.reserveCapacity(capacity)
        self.capacity = capacity
    }

    public mutating func enqueue(_ element: T) {
        if count == capacity {
            increaseCapacity()
        }

        let enqueueIndex = (firstIndex + count) % capacity
        if enqueueIndex >= array.count {
            array.append(element)
        } else {
            array[enqueueIndex] = element
        }

        count += 1
    }

    public mutating func dequeue() -> T {
        let first = array[firstIndex]
        firstIndex = (firstIndex + 1) % capacity
        count -= 1
        return first
    }

    mutating func increaseCapacity() {
        let countAtEndOfArray = capacity - firstIndex
        let countAtStartOfArray = count - countAtEndOfArray
        array.append(contentsOf: array[0..<countAtStartOfArray])
        capacity *= 2
    }
}
